server/src/main/java/org/candlepin/model/OwnerProductShareCurator.java

```diff
@@ -0,0 +1,97 @@
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.model;
+
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+
+import org.apache.commons.collections.CollectionUtils;
+
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+
+import javax.persistence.TypedQuery;
+
+/**
+ * Curator to handle creation and maintenance of OwnerProductShare objects
+ */
+public class OwnerProductShareCurator extends AbstractHibernateCurator<OwnerProductShare> {
+
+    @Inject
+    public OwnerProductShareCurator() {
+        super(OwnerProductShare.class);
+    }
+
+    public List<OwnerProductShare> findProductSharesByRecipient(Owner owner, boolean activeOnly,
+        Collection<String> productIds) {
+        return findProductSharesByOwner(owner, true, activeOnly, productIds);
+    }
+
+    public List<OwnerProductShare> findProductSharesBySharer(Owner owner, boolean activeOnly,
+        Collection<String> productIds) {
+        return findProductSharesByOwner(owner, false, activeOnly, productIds);
+    }
+
+    private List<OwnerProductShare> findProductSharesByOwner(Owner owner, boolean isRecipient,
+        boolean activeOnly, Collection<String> productIds) {
+
+        String jpql = "FROM OwnerProductShare ps WHERE ";
+        if (CollectionUtils.isNotEmpty(productIds)) {
+            jpql += "ps.productId in (:product_ids) AND ";
+        }
+
+        if (isRecipient) {
+            jpql += "ps.recipientOwner.id = :owner_id ";
+        }
+        else {
+            jpql += "ps.sharingOwner.id = :owner_id ";
+        }
+
+        if (activeOnly) {
+            jpql += "AND ps.active = true";
+        }
+
+        TypedQuery<OwnerProductShare> query = getEntityManager()
+            .createQuery(jpql, OwnerProductShare.class)
+            .setParameter("owner_id", owner.getId());
+
+        if (CollectionUtils.isNotEmpty(productIds)) {
+            query.setParameter("product_ids", productIds);
+        }
+
+        return query.getResultList();
+    }
+
+    @Transactional
+    public boolean removeShares(List<OwnerProductShare> shares) {
+        int rows = 0;
+        if (CollectionUtils.isNotEmpty(shares)) {
+            List<String> ids = new LinkedList<String>();
+            for (OwnerProductShare share : shares) {
+                ids.add(share.getId());
+            }
+
+            String jpql = "DELETE FROM OwnerProductShare s " +
+                "WHERE s.id in (:ids)";
```

Comment:
Do we need to worry about the in clause limit here?

Reply:
done